libkbfs/quota_usage.go

```diff
@@ -70,18 +77,69 @@ func NewEventuallyConsistentTeamQuotaUsage(
 	return q
 }
 
+func (q *EventuallyConsistentQuotaUsage) getCached() cachedQuotaUsage {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+	return q.cached
+}
+
+func (q *EventuallyConsistentQuotaUsage) getID(
+	ctx context.Context) (keybase1.UserOrTeamID, error) {
+	if q.tid.IsNil() {
+		session, err := q.config.KBPKI().GetCurrentSession(ctx)
+		if err != nil {
+			return keybase1.UserOrTeamID(""), err
+		}
+		return session.UID.AsUserOrTeam(), nil
+	}
+	return q.tid.AsUserOrTeam(), nil
+}
+
 func (q *EventuallyConsistentQuotaUsage) getAndCache(
 	ctx context.Context) (err error) {
 	defer func() {
 		q.log.CDebugf(ctx, "getAndCache: error=%v", err)
 	}()
+
+	// Try pulling the quota from the disk cache.  If it exists, still
+	// try the servers, but give it a short timeout.
+	var quotaInfoFromCache *kbfsblock.QuotaInfo
+	id, err := q.getID(ctx)
+	if err != nil {
+		return err
+	}
+	getCtx := ctx
+	dqc := q.config.DiskQuotaCache()
+	if dqc != nil {
+		qi, err := dqc.Get(ctx, id)
+		if err == nil {
+			q.log.CDebugf(ctx, "Read quota for %s from disk cache", id)
+			quotaInfoFromCache = &qi
+			var cancel context.CancelFunc
+			getCtx, cancel = context.WithTimeout(
+				ctx, quotaServerTimeoutWhenCached)
+			defer cancel()
```

Comment:
500ms is pretty short. Would it be better to let it keep getting it in background, until a longer timeout?

Reply:
I couldn't figure out an easy way to keep the original request going after a timeout, so instead commit 96453b4 launches a new background fetch with no timeout and updates the cache if it returns safely.  Please give it a review!